File: mobile_client/lib/models/game.dart
import 'package:mobile_client/models/question.dart';

class Game {
  final String id;
  final String title;
  final String description;
  final int duration;
  final DateTime lastModification;
  final List<Question> questions;
  final bool isHidden;

  Game(
      {required this.id,
      required this.title,
      required this.description,
      required this.duration,
      required this.lastModification,
      required this.questions,
      required this.isHidden});

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      duration: json['duration'],
      lastModification: DateTime.parse(json['lastModification']),
      questions:
          (json['questions'] as List).map((q) => Question.fromJson(q)).toList(),
      isHidden: json['isHidden'],
    );
  }
}

File: server/app/model/database/game.ts
import { removeUnderscoreId } from '@app/utils/transform-utils';
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { ApiProperty } from '@nestjs/swagger';
import { Document } from 'mongoose';
import { Question } from './question';

export type GameDocument = Game & Document;

@Schema({
    toJSON: {
        virtuals: true,
        versionKey: false,
        transform: removeUnderscoreId,
    },
    timestamps: {
        updatedAt: 'lastModification',
    },
    toObject: {
        virtuals: true,
    },
})
export class Game {
    @ApiProperty({ name: 'id' })
    _id?: string;

    @ApiProperty()
    @Prop({ required: true })
    title: string;

    @ApiProperty()
    @Prop({ required: true })
    description: string;

    @ApiProperty()
    @Prop({ required: true })
    duration: number;

    @ApiProperty({ type: [Question] })
    @Prop({ required: true })
    questions: Question[];

    @ApiProperty()
    @Prop({ required: true })
    private: boolean;

    @ApiProperty()
    @Prop({ required: true })
    creator: string;

    @ApiProperty()
    @Prop({ default: true })
    isHidden: boolean;

    @ApiProperty()
    @Prop()
    lastModification: Date;

    id: string | null;
}

export const gameSchema = SchemaFactory.createForClass(Game);
